### PR TITLE
[BI-979] Initial commit

### PR DESCRIPTION
**Pull request by henrikgrundmann**
Transfer customer_lifetime_value export to export service
Transfer rebuy_rate export to export service
Check which days to be reported on are still missing from the database
Compute values for those days
Write the results to DB, day by day
If table Does not exist in DB, create it accoring to specified sql

### CODE REVIEW

File: algorithms/customer_lifetime_value_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    df['accumulated_gmv'] = df['price_to_pay'].astype(float) * df['quantity'].astype(int)
+
+    #calculate the total gmv for each day for the customers (for the case of 
+    #several items or several sales on per day
+    df = df.groupby(['date', 'email'])['accumulated_gmv'].sum().reset_index()
+    #calculate the month of first contact
+    df['month_of_first_contact'] = df.groupby('email')['date'].transform(min).apply(lambda x:x [:7])
+    df = df.groupby(by=['date', 'month_of_first_contact']).sum().groupby(level=[0]).cumsum().reset_index()
+
+    return df[df['date'] == date]
```
Comment (tsarouch) [inline algorithms/customer_lifetime_value_algorithm.py]: just add a line in the end of the code here (to be compatible with pep8)

File: aux/datetools.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    :shifted_n_days: The number of dates of shift
+    :shift_in_future: Bollean flag, if True/False we shift in future/past
+    :date_format: The string format of the date
+    """
+    datetime_object = datetime.datetime.strptime(date_string, date_format)
+    if shift_in_future:
+        shifted_date = datetime_object + datetime.timedelta(days=shifted_n_days)
+    else:
+        shifted_date = datetime_object - datetime.timedelta(days=shifted_n_days)
+    return datetime.datetime.strftime(shifted_date, date_format)
```
Comment (tsarouch) [inline aux/datetools.py]: just add a line in the end of the code here (to be compatible with pep8)

File: scheduled_tasks/customer_livetime_value.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        rebuy_rate_io.write_result(date, VERSION, daily_clv_df, 'customer_lifetime_value')
+
+    
+final_task = PythonOperator(
+    task_id='customer_lifetime_value',
+    python_callable=upload_customer_lifetime_value,
+    retry_delay=datetime.timedelta(seconds=5),
+    retries=3,
+    trigger_rule="all_done",
+    dag=dag)
```
Comment (tsarouch) [inline scheduled_tasks/customer_livetime_value.py]: just add a line in the end of the code here (to be compatible with pep8)

Comment (tsarouch) [inline scheduled_tasks/customer_livetime_value.py]: Also is it really 
customer_livetime_value.py 
not 
customer_lifetime_value.py 
?

File: scheduled_tasks/rebuy_rate.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        rebuy_rate_io.write_result(date, VERSION, rebuy_rate_series, 'rebuy_rate')
+
+    
+final_task = PythonOperator(
+    task_id='upload_clv',
+    python_callable=rebuy_rate_calculate_and_export,
+    retry_delay=datetime.timedelta(seconds=5),
+    retries=3,
+    trigger_rule="all_done",
+    dag=dag)
```
Comment (tsarouch) [inline scheduled_tasks/rebuy_rate.py]: just add a line in the end of the code here (to be compatible with pep8)

File: scheduled_tasks/test_dag.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    dag=dag)
+
+# Generate 10 sleeping tasks, sleeping from 0 to 9 seconds respectively
+task = PythonOperator(
+    task_id='sleep',
+    python_callable=my_sleeping_function,
+    op_kwargs={'random_base': float(10) / 10},
+    dag=dag)
+
+task.set_upstream(run_this)
```
Comment (tsarouch) [inline scheduled_tasks/test_dag.py]: just add a line in the end of the code here (to be compatible with pep8)

**Comment by tsarouch** [comment]
I have something that puzzles me with this new repo.
We ll need common code, like started here:
e.g. datetools.py 
This is also in our DS repo.
How are we going to handle this ?
Maybe @carolinux @sergherrero any previous experience and good practices ?

**Comment by carolinux** [comment]
@tsarouch We will delete the rebuy rate and other code from the main DS repo when the work here is done. Now, regarding datetools, since it seems to me it's just one file we can copy it over.

Another option would be to have this repository depend on the DS repository (at a specific version/git commit), so in the machine we have it running, we also git clone the DS repository and add the directory to the `PYTHONPATH`. Then the import  statement `from aux.datetools import foo` would work just fine. I would insist on this in cases where we end up using lots of modules from the DS repo. @sergherrero feel free to weigh in

File: algorithms/rebuy_rates_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+                         .apply(set)[UNIQUE_USER_IDENTIFIER])
+    output_d['total_buyers'] = n_total_buyers
+
+    for window_n_days in WINDOW_LAG_DAYS:
+        output_d['total_buyers_repeated_within_' + \
+                 str(window_n_days) + '_days'] = \
+            get_window_repeated_buyers(
+                df, current_date, current_buyers_set, window_n_days)
+
+    return pd.Series(output_d)
```
Comment (carolinux) [inline algorithms/rebuy_rates_algorithm.py]: There are some tests in the DS repo for the rebuy rate algorithm. I will copy them over.

File: aux/rebuy_rate_io.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+#def get_session():
+#    if 'Session' not in globals():
+#        global Session
+#        engine = create_engine(os.environ["DB_URL_DATA_PUDDLE"])
+#        Session = scoped_session(sessionmaker(bind=engine))
+#    return Session()
+
+
+def get_connection():
+    if 'data_puddle_engine' not in globals():
```
Comment (carolinux) [inline aux/rebuy_rate_io.py]: data puddle :100:

File: aux/rebuy_rate_io.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    try:
+        float(value)
+        return True
+    except:
+        return False
+        
+def sql_from_file(filepath):
+    """Returns a list of sql-commands that are stored in filepath
+    """
+    f = open(filepath)
```
Comment (carolinux) [inline aux/rebuy_rate_io.py]: More pythonic to do it like this (and then don't need to close manually)
```
with open(filepath, 'r') as f:
    sql_commands = [command.strip() for command in f.read().split(';')]
```

File: aux/rebuy_rate_io.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    sql_commands = [command.strip() for command in f.read().split(';')]
+    f.close()
+    return sql_commands
+
+def check_create_table(table_name):
```
Comment (carolinux) [inline aux/rebuy_rate_io.py]: - I would rename this to `create_table_if_not_exists`
- Use the method  `sql_from_file` to read the table creation statement

File: scheduled_tasks/test_dag.py (hunk 5 above)
Comment (carolinux) [inline scheduled_tasks/test_dag.py]: we probably don't need to check this test file into source control?

File: algorithms/customer_lifetime_value_algorithm.py (hunk 1 above)
Comment (henrikgrundmann) [inline algorithms/customer_lifetime_value_algorithm.py]: done

File: aux/datetools.py (hunk 2 above)
Comment (henrikgrundmann) [inline aux/datetools.py]: done

File: scheduled_tasks/customer_livetime_value.py (hunk 3 above)
Comment (henrikgrundmann) [inline scheduled_tasks/customer_livetime_value.py]: done
and changed to _customer_lifetime_value.py_

File: scheduled_tasks/rebuy_rate.py (hunk 4 above)
Comment (henrikgrundmann) [inline scheduled_tasks/rebuy_rate.py]: done

File: scheduled_tasks/test_dag.py (hunk 5 above)
Comment (henrikgrundmann) [inline scheduled_tasks/test_dag.py]: slipped through when committing, sorry.

File: aux/rebuy_rate_io.py (hunk 8 above)
Comment (henrikgrundmann) [inline aux/rebuy_rate_io.py]: done

File: aux/rebuy_rate_io.py (hunk 9 above)
Comment (henrikgrundmann) [inline aux/rebuy_rate_io.py]: done

File: scheduled_tasks/customer_livetime_value.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import datetime
+import os
+
+import pandas as pd
+import sqlalchemy
+
+from aux import rebuy_rate_io
+from algorithms.customer_lifetime_value_algorithm import get_daily_customer_value
+
+VERSION = '1.5'
```
Comment (carolinux) [inline scheduled_tasks/customer_livetime_value.py]: I guess the version here doesn't need to be the same as the version of the rebuy rate algorithm

Comment (henrikgrundmann) [inline scheduled_tasks/customer_livetime_value.py]: True, changed.

File: scheduled_tasks/test_dag.py (hunk 5 above)
Comment (henrikgrundmann) [inline scheduled_tasks/test_dag.py]: removed from branch

**Pull request closed without merging**